sling/nlp/document/lexical-encoder.cc

```diff
@@ -0,0 +1,398 @@
+    memcpy(f, reader.embedding().data(), rowsize);
+    found++;
+  }
+  return found;
+};
+
+void LexicalFeatureExtractor::Compute(const DocumentFeatures &features,
+                                      int index, float *fv) {
+  // Extract word feature.
+  if (lex_.word_feature_) {
+    *data_.Get<int>(lex_.word_feature_) = features.word(index);
+  }
+
+  // Extract prefix feature.
+  if (lex_.prefix_feature_) {
+    Affix *affix = features.prefix(index);
+    int *a = data_.Get<int>(lex_.prefix_feature_);
+    for (int n = 0; n < lex_.prefix_size_; ++n) {
+      if (affix != nullptr) {
+        *a++ = affix->id();
+        affix = affix->shorter();
+      } else {
+        *a++ = -1;
+      }
+    }
+  }
+
+  // Extract suffix feature.
+  if (lex_.suffix_feature_) {
+    Affix *affix = features.suffix(index);
+    int *a = data_.Get<int>(lex_.suffix_feature_);
+    for (int n = 0; n < lex_.suffix_size_; ++n) {
+      if (affix != nullptr) {
+        *a++ = affix->id();
+        affix = affix->shorter();
+      } else {
+        *a++ = -1;
+      }
+    }
+  }
+
+  // Extract hyphen feature.
+  if (lex_.hyphen_feature_) {
+    *data_.Get<int>(lex_.hyphen_feature_) = features.hyphen(index);
+  }
+
+  // Extract capitalization feature.
+  if (lex_.caps_feature_) {
+    *data_.Get<int>(lex_.caps_feature_) = features.capitalization(index);
+  }
+
+  // Extract punctuation feature.
+  if (lex_.punct_feature_) {
+    *data_.Get<int>(lex_.punct_feature_) = features.punctuation(index);
+  }
+
+  // Extract quote feature.
+  if (lex_.quote_feature_) {
+    *data_.Get<int>(lex_.quote_feature_) = features.quote(index);
+  }
+
+  // Extract digit feature.
+  if (lex_.digit_feature_) {
+    *data_.Get<int>(lex_.digit_feature_) = features.digit(index);
+  }
+
+  // Set reference to output feature vector.
+  data_.SetReference(lex_.feature_vector_, fv);
+
+  // Map features through embeddings.
+  data_.Compute();
+}
+
+void LexicalFeatureExtractor::Extract(const Document &document,
+                                      int begin, int end, Channel *fv) {
+  // Extract lexical features from document.
+  DocumentFeatures features(&lex_.lexicon_);
+  features.Extract(document, begin, end);
+
+  // Compute feature vectors.
+  int length = end - begin;
+  fv->resize(length);
+  for (int i = 0; i < length; ++i) {
+    float *f = reinterpret_cast<float *>(fv->at(i));
+    Compute(features, i, f);
+  }
+}
+
+Channel *LexicalFeatureLearner::Extract(const Document &document,
+                                        int begin, int end) {
+  // Clear previous feature vectors.
+  for (auto *e : extractors_) delete e;
+  extractors_.clear();
+
+  // Extract lexical features from document.
+  DocumentFeatures features(&lex_.lexicon_);
+  features.Extract(document, begin, end);
+
+  // Compute feature vector for all tokens in range.
+  int length = end - begin;
+  fv_.resize(length);
+  for (int i = 0; i < length; ++i) {
+    auto *e = new LexicalFeatureExtractor(lex_);
```

Comment:
This repeated allocation of LexicalFeatureExtractors is only for gradient propagation, right? IIRC, for inference, in theory we can reuse a single LexicalFeatureExtractor as long as we zero-out stale feature ids.

Reply:
Yes, this is only needed during learning. We could consider reusing these so we just keep extending the extractors_ array instead of reallocating this for each sentence.